src/Microsoft.ML.PCA/PcaTransform.cs

```diff
@@ -493,56 +530,134 @@ private Float[][] PostProcess(Float[][] y, Float[] sigma, Float[] z, int d, int
-            return del;
-        }
+                    _input = columnPair.input;
+                    _output = columnPair.output;
+                    _weightColumn = weightColumn;
+                    _schema = schema;
+                }
 
-        private static void TransformFeatures(IExceptionContext ectx, ref VBuffer<Float> src, ref VBuffer<Float> dst, TransformInfo transformInfo)
-        {
-            ectx.Check(src.Length == transformInfo.Dimension);
+                public int InputIndex
+                {
+                    get
+                    {
+                        // Column names are already checked by PcaTransform
+                        _schema.TryGetColumnIndex(_input, out int index);
+                        return index;
+                    }
+                }
+
+                public ColumnType InputType => _schema[_input].Type;
+
+                public int WeightColumnIndex
+                {
+                    get
+                    {
+                        var index = -1;
+                        if (_weightColumn != null)
+                        {
+                            if (!_schema.TryGetColumnIndex(_weightColumn, out index))
+                                throw Contracts.Except("Weight column '{0}' does not exist.", _weightColumn);
+                            Contracts.CheckUserArg(_schema[index].Type == NumberType.Float, nameof(_weightColumn));
+                        }
+                        return index;
+                    }
+                }
+            }
 
-            var values = dst.Values;
-            if (Utils.Size(values) < transformInfo.Rank)
-                values = new Float[transformInfo.Rank];
+            private readonly PcaTransform _parent;
+            private readonly int _numColumns;
 
-            for (int i = 0; i < transformInfo.Rank; i++)
+            public Mapper(PcaTransform parent, Schema inputSchema)
+               : base(parent.Host.Register(nameof(Mapper)), parent, inputSchema)
             {
-                values[i] = VectorUtils.DotProductWithOffset(transformInfo.Eigenvectors[i], 0, ref src) -
-                    (transformInfo.MeanProjected == null ? 0 : transformInfo.MeanProjected[i]);
+                _parent = parent;
+                _numColumns = parent._numColumns;
+                for (int i = 0; i < _numColumns; i++)
+                {
+                    var colPair = _parent.ColumnPairs[i];
+                    var colSchemaInfo = new ColumnSchemaInfo(colPair, inputSchema);
+                    ValidatePcaInput(Host, colPair.input, colSchemaInfo.InputType);
+                    if (colSchemaInfo.InputType.VectorSize != _parent._transformInfos[i].Dimension)
+                    {
+                        var msg = $"Dimension of column ${colPair.input} is ${colSchemaInfo.InputType.VectorSize}, which doesn't match the expected size ${_parent._transformInfos[i].Dimension}";
+                        throw Host.Except(msg);
```

Comment:
>Except [](start = 35, length = 6)

this message also should be `ExceptSchemaMismatch`

Reply:
`Schema mismatch for input column 'Foo': expected V<R4, 4>, got V<R4, 5>`
Produced by `ExceptSchemaMismatch("input", colPair.input, new VectorType(NumberType.R4, _parent._transformInfos[i].Dimension).ToString(),  colSchemaInfo.InputType.ToString())`

---
In reply to: [227631956](https://github.com/dotnet/machinelearning/pull/1333#discussion_r227631956) [](ancestors = 227631956,227591272)